File: LeetCode/0001_Two_Sum.java
// Approach 1: Brute Force
// Approach 2: Two-pass Hash Table
// Approach 3: One-pass Hash Table

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for(int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if(map.containsKey(complement)) {
                return new int[] { map.get(complement), i};
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");

    }
}

File: LeetCode/0001_Two_Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        unordered_map<int,int> hashTable;

        for(int i = 0; i < nums.size(); i++) {
            int numberToFind = target-nums[i];
            if(hashTable.find(numberToFind) != hashTable.end()) {
                result.push_back(i);
                result.push_back(hashTable[numberToFind]);
                return result;
            }

            hashTable[nums[i]] = i;
        }


        return result;
    }
};
